mmedit/models/editors/guided_diffusion/adm.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from copy import deepcopy
+from typing import List, Optional
+
+import mmengine
+import torch
+from mmengine import MessageHub
+from mmengine.model import BaseModel, is_model_wrapper
+from mmengine.optim import OptimWrapperDict
+from mmengine.runner.checkpoint import _load_checkpoint_with_prefix
+from tqdm import tqdm
+
+from mmedit.registry import DIFFUSERS, MODELS, MODULES
+from mmedit.structures import EditDataSample, PixelData
+from mmedit.utils.typing import ForwardInputs, SampleList
+
+def classifier_grad(x, t, y=None, classifier_scale=1.0):
+    assert y is not None
+    with torch.enable_grad():
+        x_in = x.detach().requires_grad_(True)
+        logits = classifier(x_in, t)
+        log_probs = F.log_softmax(logits, dim=-1)
+        selected = log_probs[range(len(logits)), y.view(-1)]
+        return torch.autograd.grad(selected.sum(), x_in)[0] * classifier_scale
+
+@MODELS.register_module('ADM')
+@MODELS.register_module('GuidedDiffusion')
+@MODELS.register_module()
+class AblatedDiffusionModel(BaseModel):
+    """_summary_
+
+    Args:
+        data_preprocessor (_type_): _description_
+        unet (_type_): _description_
+        diffuser (_type_): _description_
+        use_fp16 (bool, optional): _description_. Defaults to False.
+        classifier (_type_, optional): _description_. Defaults to None.
+        pretrained_cfgs (_type_, optional): _description_. Defaults to None.
+    """
+
+    def __init__(self,
+                 data_preprocessor,
+                 unet,
+                 diffuser,
+                 use_fp16=False,
+                 classifier=None,
+                 classifier_scale=1.0,
+                 pretrained_cfgs=None):
+
+        super().__init__(data_preprocessor=data_preprocessor)
+        self.unet = MODULES.build(unet)
+        self.diffuser = DIFFUSERS.build(diffuser)
+        if classifier:
+            self.classifier = MODULES.build(unet)
+        else:
+            self.classifier = None
+        self.classifier_scale = classifier_scale
+
+        if pretrained_cfgs:
+            self.load_pretrained_models(pretrained_cfgs)
+        if use_fp16:
+            mmengine.print_log('Convert unet modules to floatpoint16')
+            self.unet.convert_to_fp16()
+
+    def load_pretrained_models(self, pretrained_cfgs):
```

Comment:
Maybe we shoud remove this function and align the implement with `MMEngine.Runner`.

Reply:
This function is used to load pre-trained module(BaseModule) weights from different source models(BaseModel or external weights). ``MMEngine.Runner`` don't support this.